src/components/Backup.tsx

```diff
@@ -90,24 +117,64 @@ export class Backup extends React.PureComponent<Props, State> {
             return this.state.serializedAppState!;
         }
         const serializedAppState = await getSerializedAppState();
-        this.setState({
+        await this.completeSetState({
             serializedAppState,
         });
         return serializedAppState;
     }
 
+    private getBackupData = (): HexString => {
+        const backupData = `${this.state.contentHash}${this.state.randomSecret}` as HexString;
+        return backupData;
+    }
+
+    private getBackupText = () => {
+        const backupText = `
+Random secret: ${this.state.randomSecret}
+Content hash: ${this.state.contentHash}
+Backup link: ${this.state.backupData}
+`;
+        return backupText;
+    }
+
+    private calculateEncryptedBackupData = async (): Promise<HexString> => {
+        const backupData = this.getBackupData();
+        if (this.state.backupPassword !== '') {
+            const plainData = new Uint8Array(hexToByteArray(backupData));
+            const backupPasswordByteArray = stringToByteArray(this.state.backupPassword);
+            const encryptedBackupData = await encrypt(plainData, backupPasswordByteArray);
+            const encryptedHexBackup = byteArrayToHex(encryptedBackupData, false) as HexString;
+            return encryptedHexBackup;
+        }
+        return backupData;
+    }
+
     private onBackupData = async () => {
         try {
-            const secretHex = stringToHex(this.state.secretText);
-            const serializedAppState = await this.getOrLoadSerializedAppState();
-            const emailBody = await createBackupFromString(serializedAppState, secretHex);
-            Debug.log('sendBackup body', emailBody);
-
-            this.showShareDialog(emailBody);
+            const contentHash = await this.backupAppStateToSwarm(this.state.randomSecret);
+            await this.completeSetState({
+                contentHash,
+            });
+            const backupData = await this.calculateEncryptedBackupData();
+            await this.completeSetState({
+                backupData,
+            });
+            Clipboard.setString(backupData);
+            await this.showShareDialog(backupData);
         } catch (e) {
             Debug.log('sendBackup error', e);
         }
     }
+
+    private backupAppStateToSwarm = async (secretHex: HexString): Promise<HexString> => {
```

Comment:
I would not put these kind of loading state information in AppState unless it has to work across several screens. Instead it would be better to store this in a state variable, either as a class member or with hooks.